modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/serviceextensions/preupdatepassword/PreUpdatePasswordActionSmsOtpFailureTestCase.java

```diff
@@ -0,0 +1,290 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.identity.integration.test.serviceextensions.preupdatepassword;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.apache.commons.lang.StringUtils;
+import org.testng.annotations.*;
+import org.wso2.carbon.automation.engine.context.TestUserMode;
+import org.wso2.identity.integration.test.base.MockSMSProvider;
+import org.wso2.identity.integration.test.recovery.model.v2.ConfirmModel;
+import org.wso2.identity.integration.test.recovery.model.v2.InitModel;
+import org.wso2.identity.integration.test.recovery.model.v2.RecoverModel;
+import org.wso2.identity.integration.test.recovery.model.v2.ResetModel;
+import org.wso2.identity.integration.test.recovery.model.v2.UserClaim;
+import org.wso2.identity.integration.test.rest.api.server.identity.governance.v1.dto.ConnectorsPatchReq;
+import org.wso2.identity.integration.test.rest.api.server.identity.governance.v1.dto.PropertyReq;
+import org.wso2.identity.integration.test.rest.api.server.notification.sender.v1.model.Properties;
+import org.wso2.identity.integration.test.rest.api.server.notification.sender.v1.model.SMSSender;
+import org.wso2.identity.integration.test.restclients.NotificationSenderRestClient;
+import org.wso2.identity.integration.test.restclients.PasswordRecoveryV2RestClient;
+import org.wso2.identity.integration.test.serviceextensions.dataprovider.model.ActionResponse;
+import org.wso2.identity.integration.test.serviceextensions.dataprovider.model.ExpectedPasswordUpdateResponse;
+import org.wso2.identity.integration.test.serviceextensions.mockservices.ServiceExtensionMockServer;
+import org.wso2.identity.integration.test.serviceextensions.model.*;
+import org.wso2.identity.integration.test.rest.api.server.notification.sender.v1.model.Error;
+import org.wso2.identity.integration.test.rest.api.user.common.model.*;
+import org.wso2.identity.integration.test.restclients.SCIM2RestClient;
+import org.wso2.identity.integration.test.utils.FileUtils;
+
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.servlet.http.HttpServletResponse;
+
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertNotNull;
+import static org.testng.Assert.assertTrue;
+
+/**
+ * Integration test class for testing the pre update password action failure.
+ * This test case extends {@link PreUpdatePasswordActionBaseTestCase} and focuses on failed scenario
+ * on SMS OTP based password recovery/reset via V2 recovery API.
+ */
+public class PreUpdatePasswordActionSmsOtpFailureTestCase extends PreUpdatePasswordActionBaseTestCase {
```

Comment:
Shall we have this inside the `PreUpdatePasswordActionFailureTestCase` class, without having a separate class.
Then we can run for Error scenario as well.
Can reduce duplicated codes. 

We can add the common methods such as (initSmsOtpserver, updatePasswordBasedRecoveryFeatureStatus, exactOtp) to the `PreUpdatePasswordActionBaseTestCase`

Reply:
Since this test targets only the V2 API flow, the current logic feels unnecessarily complex. Additionally, it could be enhanced to support the emails-v2 flow as well for better coverage.